p2p/src/utils/requester.rs

```diff
@@ -78,29 +92,28 @@ impl<E: Clock + GClock, C: Scheme> Requester<E, C> {
         self.rate_limiter.shrink_to_fit();
     }
 
-    /// Once a participant is added to skip, it is never removed.
-    pub fn skip(&mut self, participant: PublicKey) {
+    /// Skip a participant for future requests.
+    ///
+    /// Participants added to this list will never be removed (even if dropped
+    /// during `reconcile`, in case they are re-added later).
+    pub fn block(&mut self, participant: PublicKey) {
         self.excluded.insert(participant);
     }
 
-    /// Remove an outstanding request.
-    fn remove(&mut self, id: ID) -> Option<(PublicKey, SystemTime)> {
-        let (participant, start, deadline) = self.requests.remove(&id)?;
-        let ids = self
-            .deadlines
-            .get_mut(&deadline)
-            .expect("deadline not found");
-        ids.remove(&id);
-        if ids.is_empty() {
-            self.deadlines.remove(&deadline);
+    /// Ask for a participant to handle a request.
+    ///
+    /// If `shuffle` is true, the order of participants is shuffled before
+    /// a request is made. This is typically used when a request to the preferred
+    /// participant fails.
+    pub fn request(&mut self, shuffle: bool) -> Option<(PublicKey, ID)> {
+        // Shuffle participants
+        let mut participants = self.participants.iter().collect::<Vec<_>>();
+        if shuffle {
+            participants.shuffle(&mut self.runtime);
         }
-        Some((participant, start))
-    }
 
-    /// Ask for a participant to handle a request.
-    pub fn request(&mut self) -> Option<(PublicKey, ID)> {
         // Look for a participant that can handle request
-        for (participant, _) in self.participants.iter() {
+        for (participant, _) in participants {
```

Comment:
You may have realized this, but this is less efficient when not sorting. Is it possible somehow to keep it as an iterator? Something like:

```
let mut participants = self.participants.iter();
if shuffle {
  participants = participants.collect::<Vec<_>>().shuffle(&mut self.runtime).iter();
}

for (participant, _) in participants {
  // ...
```

Reply:
I took an initial stab at this (for the same reason you called out) but gave up because it wasn't clean.

Dug deeper this time and got it to compile with:
```
        // Shuffle participants
        let participants = self.participants.iter();
        let participants = if !shuffle {
            Box::new(participants) as Box<dyn std::iter::Iterator<Item = (&Bytes, &u128)>>
        } else {
            let mut participants = participants.collect::<Vec<_>>();
            participants.shuffle(&mut self.runtime);
            Box::new(participants.into_iter())
        };

```

Will think about it some more.

Reply:
You could avoid dynamic dispatch with something like this:
```
enum ParticipantIter<'a> {
    Unshuffled(std::slice::Iter<'a, (&Bytes, &u128)>),
    Shuffled(std::vec::IntoIter<(&'a Bytes, &'a u128)>),
}

impl<'a> Iterator for ParticipantIter<'a> {
    type Item = (&'a Bytes, &'a u128);

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            ParticipantIter::Unshuffled(iter) => iter.next(),
            ParticipantIter::Shuffled(iter) => iter.next(),
        }
    }
}

pub fn request(&mut self, shuffle: bool) -> Option<(PublicKey, ID)> {
    let participants = if !shuffle {
        ParticipantIter::Unshuffled(self.participants.iter())
    } else {
        let mut shuffled = self.participants.iter().collect::<Vec<_>>();
        shuffled.shuffle(&mut self.runtime);
        ParticipantIter::Shuffled(shuffled.into_iter())
    };
    .....
```

Reply:
Oooooo found a clean way using: https://crates.io/crates/either